src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -970,53 +970,29 @@ private AttributeSyntax ParseAttribute()
                 return null;
 
             var openParen = this.EatToken(SyntaxKind.OpenParenToken);
-            var argNodes = _pool.AllocateSeparated<AttributeArgumentSyntax>();
 
-tryAgain:
-            if (this.CurrentToken.Kind != SyntaxKind.CloseParenToken)
-            {
-                if (this.IsPossibleAttributeArgument() || this.CurrentToken.Kind == SyntaxKind.CommaToken)
-                {
-                    // first argument
-                    argNodes.Add(this.ParseAttributeArgument());
-
-                    // comma + argument or end?
-                    int lastTokenPosition = -1;
-                    while (IsMakingProgress(ref lastTokenPosition))
-                    {
-                        if (this.CurrentToken.Kind == SyntaxKind.CloseParenToken)
-                        {
-                            break;
-                        }
-                        else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleAttributeArgument())
-                        {
-                            argNodes.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
-                            argNodes.Add(this.ParseAttributeArgument());
-                        }
-                        else if (this.SkipBadAttributeArgumentTokens(ref openParen, argNodes, SyntaxKind.CommaToken) == PostSkipAction.Abort)
-                        {
-                            break;
-                        }
-                    }
-                }
-                else if (this.SkipBadAttributeArgumentTokens(ref openParen, argNodes, SyntaxKind.IdentifierToken) == PostSkipAction.Continue)
-                {
-                    goto tryAgain;
-                }
-            }
+            var argNodes = this.ParseCommaSeparatedSyntaxList(
+                ref openParen,
+                SyntaxKind.CloseParenToken,
+                static @this => @this.IsPossibleAttributeArgument(),
+                static @this => @this.ParseAttributeArgument(),
+                static (@this, openParen, argNodes, kind, _) => skipBadAttributeArgumentTokens(@this, openParen, argNodes, kind),
+                allowTrailingSeparator: false);
```

Comment:
this is the simplest form of calling into the helper.  You pass hte open token in (so errors can be attached to it if necessary), the token that finishes the list, the check if we're on a list element, the way to parse the list element, the function to determine skip/abort behavior, and if trailing separators are legal or not.

Reply:
the general form of this callback changed from passing the first arg by-ref, to taking it in as an arg and passing out as a return value.  This was because it needs to be callable from a lambda, and lambdas can't have `ref` parameters *unless* they also spell out the types of hte parameters.  That woudl make the syntax of hte 3rd lambda *horrendous*.  So, instead, i made it a normal parameter/return-value to make it possible to easily write these static lambdas.